packages/react-async/src/tests/component.test.tsx

```diff
@@ -1,176 +1,872 @@
+      });
+
+      it('marks the scripts as being used on the current page, and styles as being used immediately', async () => {
+        const id = random.uuid();
+        const AsyncComponent = createAsyncComponent({
+          id: () => id,
+          load: () => createResolvablePromise(ResolvedComponent).promise,
+          deferHydration: DeferTiming.Mount,
+        });
+
+        expect(
+          await getUsedAssets(<AsyncComponent />, AssetTiming.CurrentPage),
+        ).toContainEqual({id, scripts: true, styles: false});
+
+        expect(
+          await getUsedAssets(<AsyncComponent />, AssetTiming.Immediate),
+        ).toContainEqual({id, scripts: false, styles: true});
+      });
+    });
+
+    describe('DeferTiming.Idle', () => {
+      it('starts loading the component on idle callback', () => {
+        const load = jest.fn(
+          () => createResolvablePromise(ResolvedComponent).promise,
+        );
+        const AsyncComponent = createAsyncComponent({
+          load,
+          deferHydration: DeferTiming.Idle,
+        });
+        const asyncComponent = mount(<AsyncComponent />);
+
+        expect(load).not.toHaveBeenCalled();
+
+        asyncComponent.act(() => {
+          requestIdleCallback.runIdleCallbacks();
+        });
+
+        expect(load).toHaveBeenCalled();
+      });
+
+      it('marks the scripts as being used on the current page, and styles as being used immediately', async () => {
+        const id = random.uuid();
+        const AsyncComponent = createAsyncComponent({
+          id: () => id,
+          load: () => createResolvablePromise(ResolvedComponent).promise,
+          deferHydration: DeferTiming.Idle,
+        });
+
+        expect(
+          await getUsedAssets(<AsyncComponent />, AssetTiming.CurrentPage),
+        ).toContainEqual({id, scripts: true, styles: false});
+
+        expect(
+          await getUsedAssets(<AsyncComponent />, AssetTiming.Immediate),
+        ).toContainEqual({id, scripts: false, styles: true});
+      });
+    });
+
+    describe('DeferTiming.InViewport', () => {
+      it('does not start loading the component if it is not in the viewport', () => {
+        const load = jest.fn(
+          () => createResolvablePromise(ResolvedComponent).promise,
+        );
+        const AsyncComponent = createAsyncComponent({
+          load,
+          deferHydration: DeferTiming.InViewport,
+        });
+        const asyncComponent = mount(<AsyncComponent />);
+
+        asyncComponent.act(() => {
+          intersectionObserver.simulate({
+            isIntersecting: false,
+          });
+        });
+
+        expect(load).not.toHaveBeenCalled();
+      });
+
+      it('starts loading the component when it is in the viewport', () => {
+        const load = jest.fn(
+          () => createResolvablePromise(ResolvedComponent).promise,
+        );
+        const AsyncComponent = createAsyncComponent({
+          load,
+          deferHydration: DeferTiming.InViewport,
```

Comment:
Not for this PR, but thinking about usecases for loading on scroll I wonder if we might want `DeferTiming.AlmostInViewPort` that starts loading when the screen is a certain distance away from the content? or `DeferTiming.ScrollContainer` that starts loading when the container is scrolled

Reply:
The API is pretty unworkable once you get more complex, I think cases like that would be better off using useAsync directly.

Reply:
fair